content/go/concepts/working-with-json/working-with-json.md

```diff
@@ -0,0 +1,228 @@
+
+    // List the menu of meals
+    for _, meal := range meals.Meals {
+        fmt.Printf("---\nMeal: %s\ningredients: %s \n", meal.Name, strings.Join(meal.Ingredients, ", "))
+    }
+
+    // Serialize the object to a slice of bytes
+    data, err := json.Marshal(meals)
+    if err != nil {
+        fmt.Println(err)
+    }
+    fmt.Println("-------------")
+    fmt.Println(string(data))
+
+    // Crate an empty file
+    file, err := os.Create("meals.json")
+    if err != nil {
+        fmt.Println("Error opening file:", err)
+        return
+    }
+
+    defer file.Close()
+
+    // Write data into the file previously created
+    if _, err = file.Write(data); err != nil {
+        fmt.Println(err)
+        return
+    }
+
+}
+```
+
+## NewEncoder and NewDecoder
+
+In Go `json.NewDecoder` is used to format data received from a REST API. Other function that is commonly used is `json.NewEncoder` to exchange data between backend side and external clients that require a JSON response.
+
+In the following example, we will demonstrate the usage of both functions. We will make a call to an external REST API and enhance the result by adding products made from the fruit specified in the query parameter. For example: `http://localhost:4444/fruit?name=Strawberry`
+
+```go
+package main
+
+import (
+    "encoding/json"
+    "errors"
+    "fmt"
+    "io"
+    "log"
+    "net/http"
+    "strings"
+)
+
+type ProductsByFruit map[string][]string
+
+type Fruit struct {
+    Name     string   `json:"name"`
+    ID       int      `json:"id"`
+    Family   string   `json:"family"`
+    Order    string   `json:"order"`
+    Genus    string   `json:"genus"`
+    Products []string `json:"products"`
+}
+
+func main() {
+    mux := http.NewServeMux()
+    mux.HandleFunc("POST /fruit", func(w http.ResponseWriter, r *http.Request) {
+
+        // Get query params from URL
+        params := r.URL.Query()
+
+        // We are going to validate the name of the fruit
+        if fruitName := params.Get("name"); len(fruitName) > 0 {
+
+            // Set the specific content type for the response
+            w.Header().Set("Content-Type", "application/json")
+
+            // Get the results of the invocation to the REST API.
+            result, err := GetFruitData(fruitName)
+
+            // We need to handle any errors that occur during the decoding process.
+            if err != nil {
+                http.Error(w, "Error fruit data not found", http.StatusNotFound)
+                return
+            }
+
+            // We need a variable to store the results from the decoding process.
+            fruit := Fruit{}
+
+            // To Decode the data, we need to provide:
+            // 1. Data in the format of io.ReadCloser.
+            // 2. A pointer to the location where the data will be stored.
+            err = json.NewDecoder(result).Decode(&fruit)
```

Comment:
The `GetFruitData` function doesn't check if the response is `nil` before using it. In the case of an error (e.g., network failure), response could be `nil`, leading to a runtime panic.

Fix:

```go
if response == nil {
    return nil, errors.New("no response from the server")
}
```